### PR TITLE
Java gateway bugfixes

### PR DESCRIPTION
**Pull request by dominikl**
# What this PR does
- Guest login was used to determine if a username is actually a session id. With this PR a `joinSession` attempt is made, if the username matches UUID format.
- The `Facility.handleException()` method silently 'swallowed' exceptions caused by connection issues. With this PR a `DSOutofServiceException` is thrown in these cases.
- Adds `JoinSessionCredentials` to provide a specific class for session id logins.
- Automatically remove stale connectors from the groupConnectorMap. That way the Gateway can continue working after a server restart without explicit dis- and reconnect.
- Makes the `Gateway` `AutoCloseable` so that it can be used in try-with-resources blocks.
- In various Facility methods, check arguments and return immediately when arguments not reasonable (Ids not provided, etc.).

# Testing this PR

- Check Java gateway integration tests.
- Test that login via session id still works
- Test that reconnect after connection loss still works. E. g. open Insight. Disconnect network. Try to browse a dataset. A connection warning should appear after a while. Reconnect the network. Insight should reconnect again, too, without crashing.
- Continuously request some data from the server using the same Gateway connection (catch all exceptions). Stop and restart the server in between. While the server is not running you will get some exceptions. But when the server is ready again the Gateway should continue working as usual. See below for an example test client (*).

# Related reading

- https://trello.com/c/XKg3FisH/42-java-gateway
- https://trello.com/c/JF3VjeWk/694-refactor-logincredentials
- Context for `Facility.handleException()` and server restart issue:  https://www.openmicroscopy.org/community/viewtopic.php?f=6&t=8367

--

*) Example client to use for testing server restart
```
LoginCredentials l = new LoginCredentials( ... );
Gateway g = new Gateway(new SimpleLogger());

ExperimenterData exp = g.connect(l);
SecurityContext ctx = new SecurityContext(exp.getGroupId());

while (true) {
    try {
        IObject obj = g.getFacility(BrowseFacility.class).findIObject(
                ctx, Image.class.getSimpleName(), 1);
        System.out.println(obj.getClass().getSimpleName() + " "
                + obj.getId().getValue());
    } catch (Throwable e) {
        e.printStackTrace();
    } finally {
        try {
            Thread.sleep(5000);
        } catch (InterruptedException e) {
        }
    }
}
```

### CODE REVIEW

File: components/blitz/src/omero/gateway/Gateway.java [hunk 1]
```diff
@@ -142,7 +142,7 @@
     public static final String PROP_STATELESS_SERVICE_CREATED = "PROP_STATELESS_SERVICE_CREATED";
 
     /** Regex matching ICE Session ID */
-    private static final String ICE_SESSION_ID_REGEX = "[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}";
+    public static final String ICE_SESSION_ID_REGEX = "[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}";
```
Comment (joshmoore) [inline components/blitz/src/omero/gateway/Gateway.java]: Interesting. Might be worth moving this to an Ice constant and using this server-side as well. To data, I've always treated the session IDs as opaque, i.e. no assumptions could be made.

File: components/tools/OmeroJava/test/integration/gateway/GatewayUsageTest.java [hunk 2]
```diff
@@ -73,6 +74,31 @@ public void testLoginWithArgs()
...
+        String sessionId = gw.getSessionId(root);
+
+        Gateway gw2 = new Gateway(new SimpleLogger());
+        JoinSessionCredentials c2 = new JoinSessionCredentials(sessionId,
+                client.getProperty("omero.host"), Integer.parseInt(client
+                        .getProperty("omero.port")));
+        ExperimenterData root2 = gw2.connect(c2);
+        Assert.assertNotNull(root2);
+        Assert.assertEquals(gw2.getSessionId(root2), sessionId);
+        gw2.disconnect();
```
Comment (joshmoore) [inline components/tools/OmeroJava/test/integration/gateway/GatewayUsageTest.java]: If something throws re: `gw2`, `gw1` will not be closed which could hang the test.

File: components/blitz/src/omero/gateway/Gateway.java (hunk 1 above)
Comment (dominikl) [reply components/blitz/src/omero/gateway/Gateway.java]: The documentation says that each Ice object has an UUID and so does the session ID look like, although I didn't find anything in the documentation which explicitely says that the session ID is an UUID, too.

Comment (dominikl) [reply components/blitz/src/omero/gateway/Gateway.java]: Hm, actually I could also just use Java's builtin `UUID.fromString()` method.

Comment (joshmoore) [reply components/blitz/src/omero/gateway/Gateway.java]: > although I didn't find anything in the documentation which explicitely says that the session ID is an UUID, too.

Exactly. It's not *yet* part of the contract.

File: components/tools/OmeroJava/test/integration/gateway/GatewayUsageTest.java (hunk 2 above)
Comment (dominikl) [reply components/tools/OmeroJava/test/integration/gateway/GatewayUsageTest.java]: Could make the Gateway AutoCloseable ( c31e719 ), would make things easier. Any objections?

**Comment by dominikl** [comment]
Just for reference, already noticed the problem with the 'swallowed' exceptions in 5.2 https://github.com/openmicroscopy/openmicroscopy/pull/4344#issuecomment-161947619 and Josh's idea of an 'ExceptionHandler', but forgot about it again. General exception handling overhaul would be a breaking change, but probably necessary, maybe for 5.5 then...

File: components/blitz/src/omero/gateway/facility/BrowseFacility.java [hunk 3]
```diff
@@ -130,7 +130,7 @@
     public Collection<DataObject> getHierarchy(SecurityContext ctx, Class rootType,
             List<Long> rootIDs, Parameters options)
             throws DSOutOfServiceException, DSAccessException {
-        if (rootType == null || CollectionUtils.isEmpty(rootIDs))
+        if (rootType == null)
```
Comment (mtbc) [inline components/blitz/src/omero/gateway/facility/BrowseFacility.java]: The `rootIDs.isEmpty()` case will go okay?

Comment (dominikl) [reply components/blitz/src/omero/gateway/facility/BrowseFacility.java]: Yes, same behaviour as roodIDs == null (get all).

Comment (mtbc) [reply components/blitz/src/omero/gateway/facility/BrowseFacility.java]: Aha, maybe also mention in Javadoc, just add "or empty" or something.

File: components/tools/OmeroJava/test/integration/gateway/GatewayUsageTest.java (hunk 2 above)
Comment (mtbc) [reply components/tools/OmeroJava/test/integration/gateway/GatewayUsageTest.java]: No objections. :smiley:

**Comment by dominikl** [comment]
Not quite sure, ready to merge or relist for standup?

**Comment by mtbc** [comment]
Has it yet had any functional testing?

**Comment by dominikl** [comment]
Nope, doesn't look like it. I'll relist and maybe @pwalczysko can have a look when he can spare some time.

**Comment by pwalczysko** [comment]
All the four bullet points from the "Testing this PR" header passed.
Ready to merge FMPOV

**Comment by joshmoore** [comment]
Thanks, @dominikl. At some point, it would be interesting to review your changes here against the Python Gateway (i.e. properly specifying the string format for the session ids, etc.) Merging into the `clisplit` branch for further testing.

**Pull request merged**